types/artifacts.d.ts

```diff
@@ -213,6 +215,44 @@ declare global {
         requestId: string | null
       }
 
+      /** @see https://sourcemaps.info/spec.html#h.qz3o9nc69um5 */
+      export type RawSourceMap = {
+        /** File version and must be a positive integer. */
+        version: number
+        /** A list of original source files used by the `mappings` entry. */
+        sources: string[]
+        /** A list of symbol names used by the `mappings` entry. */
+        names?: string[]
+        /** An optional source root, useful for relocating source files on a server or removing repeated values in the `sources` entry. This value is prepended to the individual entries in the `source` field. */
+        sourceRoot?: string
+        /** An optional list of source content, useful when the `source` can’t be hosted. The contents are listed in the same order as the sources in line 5. `null` may be used if some original sources should be retrieved by name. */
+        sourcesContent?: string[]
+        /** A string with the encoded mapping data. */
+        mappings: string
+        /** An optional name of the generated code (the bundled code that was the result of this build process) that this source map is associated with. */
+        file?: string
+      }
+
+      /**
+       * Source map for a given script found at scriptUrl. If there is an error in fetching or
+       * parsing the map, errorMessage will be defined instead of map.
+       */
+      export type SourceMap = {
+        /** URL of code that source map applies to. */
+        scriptUrl: string
+        /** URL of the source map. undefined if from data URI. */
+        sourceMapUrl?: string
+        /** Source map data structure. */
+        map: RawSourceMap
+      } | {
```

Comment:
Just for my own `ts` sense, is this the best way of representing this?  The overlap of `scriptUrl` and `sourceMapUrl` seems odd?  Is it not good to have map and errorMessage be optional?  or mutually exclusive somehow?

Reply:
an alternative is to define a base class with all the shared properties, and extend for each "one of" property. however doing that makes the vscode's hover type hints opaque.

```ts
interface SourceMapBase {
  /** URL of code that source map applies to. */
  scriptUrl: string
  /** URL of the source map. undefined if from data URI. */
  sourceMapUrl?: string
}
interface SourceMapWithMap {
  /** Source map data structure. */
  map: RawSourceMap
}
interface SourceMapWithError {
  /** Error that occurred during fetching or parsing of source map. */
  errorMessage: string
}
/**
 * Source map for a given script found at scriptUrl. If there is an error in fetching or
 * parsing the map, errorMessage will be defined instead of map.
 */
export type SourceMap = SourceMapWithMap | SourceMapWithError;
```
hover over `LH.Artifacts.SourceMap`:
```ts
type LH.Artifacts.SourceMap = LH.Artifacts.SourceMapWithMap | LH.Artifacts.SourceMapWithError
```

instead of:
```ts
type LH.Artifacts.SourceMap = {
    scriptUrl: string;
    sourceMapUrl?: string | undefined;
    map: LH.Artifacts.RawSourceMap;
} | {
    scriptUrl: string;
    sourceMapUrl?: string | undefined;
    errorMessage: string;
}
```

the former is more useful to get the idea immediately, but it doesn't tell you what the properties are ..

Reply:
Hm, very cool, TIL and thanks for the detailed write-up.  But I think this only confirms my bias that typescript should be illegal.

Reply:
@brendankenny wdyt

Reply:
agree with all of the above (and *definitely* with avoiding a base class when inheritance is not at all necessary :), but it is going to be awkward when using in audits. Will have to do `in` checks (e.g. `if ('errorMessage' in map) {` everywhere they're used instead of normal falsey checks (which will get you "Property 'errorMessage' does not exist on type 'SourceMap'.").

Could
- live with it (keep it contained in a `filter()` or whatever)
- make one of the variables a discriminant in a real discriminated union, e.g. add `map: undefined` to the second member of the union. Then any code just needs to check for `!sourceMap.map` and it'll know if it's in the good or error case (basically whether or not `errorMessage` will be defined)